src/charm.py

```diff
@@ -1,103 +1,292 @@
-                event.defer()
-                self.unit.status = ops.WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
+        self.framework.observe(self.database.on.database_created, self._on_database_created)
+        self.framework.observe(self.database.on.endpoints_changed, self._on_database_changed)
+
+        self.framework.observe(self.ingress.on.ready, self._on_ingress_ready)
+        self.framework.observe(self.ingress.on.revoked, self._on_ingress_revoked)
+
+        self.framework.observe(
+            self.loki_consumer.on.promtail_digest_error,
+            self._promtail_error,
+        )
 
     @property
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+    def _metrics_external_url(self) -> str:
+        return normalise_url(self.ingress.url) if self.ingress.is_ready() else ""
+
+    @property
+    def _glauth_service_is_running(self) -> bool:
+        if not self._container.can_connect():
+            return False
+
+        try:
+            service = self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return service.is_running()
+
+    @property
+    def _pebble_layer(self) -> Layer:
+        pebble_layer: LayerDict = {
+            "summary": "GLAuth Application layer",
+            "description": "pebble layer for GLAuth service",
             "services": {
-                "httpbin": {
+                self._container_name: {
                     "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
+                    "summary": "GLAuth Operator layer",
+                    "startup": "disabled",
+                    "command": '/bin/sh -c "{} 2>&1 | tee {}"'.format(
+                        self._glauth_service_command,
+                        str(self._log_path),
+                    ),
                 }
             },
         }
+        return Layer(pebble_layer)
+
+    @property
+    def _ldap_port(self) -> str:
+        return self.config["ldap_port"]
+
+    @property
+    def _http_port(self) -> str:
+        return self.config["http_port"]
+
+    @property
+    def _basedn(self) -> str:
+        # baseDN example: "dc=glauth,dc=com"
+        dn_input = self.config["base_dn"]
+        list_dns = dn_input.split(",")
+        list_dc = [f"dc={dn}" for dn in list_dns]
+        return ",".join(list_dc)
```

Comment:
Any reason we don't expect the user to define the config with the proper syntax, ie: `dc=example,dc=com`. Since it is a pretty common config for LDAP I don't think we need to be doing this transformation.

Reply:
I'm somewhat new to LDAP, but I've seen a bunch of docs where people use space instead of a coma, or use all caps for for dc, and ou and such. It would be probably better though to change it, and then normalize the input.